singularity-eos/base/math_utils.hpp

```diff
@@ -22,32 +22,40 @@ namespace math_utils {
 
 /*
  * Replaces "SQUARE, CUBE, etc
- * Generic is recursive, but we specialize a few of them strategically
- * A generalization would be to treat even powers specially
+ * Use ternary operator instead of if constexpr to enable compile-time tail-recursion
+ * for various edge cases. Following accelerants are used:
+ * x^{2*i} = x^i x^i
+ * x^i = i x^{i - 1}
+ * x^{-i} = 1/x^i
  */
-template <size_t P>
-PORTABLE_FORCEINLINE_FUNCTION constexpr auto pow(const Real x) {
-  return x * pow<P - 1>(x);
+template <int P>
+PORTABLE_FORCEINLINE_FUNCTION constexpr Real pow(const Real x) {
+  // defining Ppos is important. Otherwise compiler is not smart
+  // enough to terminate recursion.
+  constexpr int Ppos = (P < 0) ? -P : P;
+  return (P >= 0)
+             ? ((Ppos % 2) ? x * pow<Ppos - 1>(x) : pow<Ppos / 2>(x) * pow<Ppos / 2>(x))
+             : 1 / pow<Ppos>(x);
 }
```

Comment:
Very pleased with this new version of our integer power function. It's generically recursive, with the even/odd power performance trick, so we don't need the strategic template specializations anymore. And it handles negative integer powers.